avalanche/caip2.md

```diff
@@ -0,0 +1,126 @@
+---
+namespace-identifier: avax-caip2
+title: Avalanche Namespace - Chains
+author: Gergely Lovas (@gergelylovas)
+discussions-to: []
+status: Draft
+type: Standard
+created: 2024-07-16
+requires: CAIP-2
+---
+
+# CAIP-2
+
+*For context, see the [CAIP-2][] specification.*
+
+## Rationale
+
+In CAIP-2 a general blockchain identification scheme is defined. This is the
+implementation of CAIP-2 for the Avalanche Ecosystem. 
+Blockchains in the "avax" namespace are validated by their `blockchainID`. The `blockchainID` in Avalanche refers to the [txID][] that created the blockchain on the Avalanche P-Chain.
+These blockchain IDs require transformations to be used as conformant CAIP-2
+references.
+
+## Syntax
+
+The namespace "avax" refers to the Avalanche open-source blockchain platform.
+
+### Reference Definition
+
+The definition for this namespace will use the `blockchainID` as an identifier
+for different Avalanche chains. Since the P-Chain's `blockchainID` is the same on Testnet and Mainnet, to ensure uniqueness, a testnet prefix is introduced for chains on the [Fuji Testnet][].  
+
+The method for calculating the chain ID is as follows with pseudo-code:
+
+```
+first_32_chars(hex(sha256(concat(testnet_prefix, blockchain_id))))
+```
+
+- `blockchain_id` = `txID` that created the blockchain on the Avalanche P-Chain
+- `testnet_prefix`= `fuji` when the chain was created on the Fuji Testnet; empty string otherwise
+- `concat`= a string concatenation function
+- `sha256`= a SHA256 hash function
+- `hex`= a lowercase hex encoder
+- `first_32_chars`= a function to extract the first 32 characters of the
+  resulting string and dropping the rest
+
+### Resolution Method
+
+To resolve a blockchain reference for the Avalanche namespace, make a JSON-RPC request to the [Info API][] with method
+`info.getBlockchainID`, for example:
+
+```jsonc
+// Request
+{
+    "jsonrpc":"2.0",
+    "id"     :1,
+    "method" :"info.getBlockchainID",
+    "params": {
+        "alias":"C"
+    }
+}
+
+// Response
+{
+  "jsonrpc": "2.0",
+  "id": 1,
+  "result": {
+    "blockchainID": "2q9e4r6Mu3U68nU1fYjgbR6JvwrRx36CohpAX5UQxse55x1Q5"
+  }
+}
+```
+
+For example, this Node.js code transforms the above response into a CAIP-2 identifier:
+
+```javascript
+const isTestnet = false;
+const blockchainID = "2q9e4r6Mu3U68nU1fYjgbR6JvwrRx36CohpAX5UQxse55x1Q5";
+const testnetPrefix = isTestnet ? 'fuji' : '';
+const hash = createHmac('sha256', testnetPrefix + blockchainID).digest('hex');
+const prefix = hash.substring(0, 32);
```

Comment:
Does this have any advantage over using hash function which can produce 16 byte hash value out of the box, for example RIPEMD-128 or similar?
Theoretically it's inevitable to have collisions  when trying to map larger string to shorter but using sha256 kinda gives false sense of uniqueness since we're slashing it's space by half on next line.

Reply:
I used SHA256 because its very common and well known. It's safe to take a substring of it and other chains like Cosmos are doing it. 
However your comment got me thinking and I ended up changing the the way we stringify from `hex` to `base64URL`. This change made the truncation unnecessary.